scene/main/node.cpp

```diff
@@ -1040,20 +1066,16 @@ String Node::_generate_serial_child_name(Node *p_child) {
 	}
 
 	String nnsep = _get_name_num_separator();
-	int num = 0;
-	bool explicit_zero = false;
-	if (nums.length() > 0 && name.substr(name.length() - nnsep.length() - nums.length(), nnsep.length()) == nnsep) {
-		// Base name + Separator + Number
-		num = nums.to_int();
-		name = name.substr(0, name.length() - nnsep.length() - nums.length()); // Keep base name
-		if (num == 0) {
-			explicit_zero = true;
-		}
+	int name_last_index = name.length() - nnsep.length() - nums.length();
+
+	// Keep base name
+	if (nums.length() > 0 && name.substr(name_last_index, nnsep.length()) == nnsep) {
+		name = name.substr(0, name_last_index).strip_edges();
 	}
 
-	int num_places = nums.length();
 	for (;;) {
-		String attempt = (name + (num > 0 || explicit_zero ? nnsep + itos(num).pad_zeros(num_places) : "")).strip_edges();
+		String attempt = name + nums;
+
 		bool found = false;
 		for (int i = 0; i < data.children.size(); i++) {
```

Comment:
Can we maybe rethink this method to avoid having a double for loop like here? Refetching all children at each attempt seems pretty inefficient to me.

A simple algorithm IMO would do:
- get base name without number
- go once over children, extract the used suffixes for children using that base name
- do the attempt/increment logic only comparing to those suffixes, without need to recheck all nodes all the time

Reply:
I've saved the children matching the base name in the variable `children_names` so we only need to use `find(attempt)` to know if the name is available.